### ISSUE TITLE
Continuous performance benchmarking

### ISSUE THREAD

**Issue by floatdrop** [sindresorhus/module-requests#24]
For some performance critical modules would be cool to track performance degradation in pull-request (with travis-ci).

__Idea__: checkout before changes, run benchmarks, checkout on head and run benchmarks again. Compare result and show warnings (or fail).

Maybe there is already a solution to this?

**Comment by sindresorhus** [comment]
I think @Qix- were working on something like this.

**Comment by Qix-** [comment]
This is exactly what I want to do.

Here are some of the difficulties as opposed to, say, coveralls.io.

- What is a *score*? What constitutes a *bad*/*fast*/*slow* program? Each program is different, and a main loop is going to take (theoretically) close to 100% of execution time.
- Benchmarking has to do with CPU clock speed, scheduling load, affinity, priority, and thanks to the stopping problem, extra random incurred costs with non-stable functions (i.e. same input, different output between runs).
- Getting some quasi-comparable value that can be contrasted to previous runs, regardless of the actual machine running them (tests run completely differently run-to-run on Travis, for example, and each run usually (always?) starts a new container instance).
- Defining a universal format that can work with all performance tools (i.e. V8's profiling dump, callgrind, etc.)

One of the things I was considering is to grade the performance of individual pieces of code to others and get a relative score (i.e. this function took 89% of the executing time), though that doesn't at all help the score problem.

**Comment by floatdrop** [comment]
There is no score, tool can just compare abstract numeric values and show, that in second benchmark run number is differs from first run. This approach will eliminate concerns about run-to-run differences, because it would be done in one run (but it will take a while to run benchmarks twice).

I often do not look how much each function runs in module, but care about performance increase/decrease in general.

@Qix-

**Comment by Qix-** [comment]
I definitely (personally) care about per-function performance, though I understand overall performance might be enough for someone.

However, I don't see how a run-to-run is going to be either helpful or reliable. Like I stated before, a program will usually never execute 1:1 at the same speed from run to run. Showing a delta between two runs isn't going to provide much information. As well, on a Travis container structure, a program will run a lot slower than on a normal computer. This will affect overall times, and due to the fact the infrastructure consists of virtual instances on a shared computer, peak usage times for Travis will translate to even slower execution times.

**Comment by floatdrop** [comment]
Understood. I tested it a little and yes - results differs from run to run, so benchmarks with slow operations will get false-negative from comparing benchmark results. But for small modules (or functions), that can do millions of op/s it can be bearable (https://travis-ci.org/floatdrop/test-bechmark/jobs/75960875):

```
small x 1,309,532 ops/sec ±0.18% (100 runs sampled)
opts x 664,409 ops/sec ±1.01% (92 runs sampled)
big x 281,449 ops/sec ±0.58% (99 runs sampled)

... tests ...

small x 1,408,360 ops/sec ±0.13% (99 runs sampled)
opts x 640,248 ops/sec ±1.45% (97 runs sampled)
big x 284,838 ops/sec ±0.30% (100 runs sampled)
```

I really want to avoid own infrastructure for such thing and use travis - but seems like it will be rather difficult (if not impossible).

> I definitely (personally) care about per-function performance, though I understand overall performance might be enough for someone.

Per-function performance can be measured by writing benchmark for this functions.

**Comment by Qix-** [comment]
> I really want to avoid own infrastructure for such thing and use travis - but seems like it will be rather difficult (if not impossible).

It would be integrated with Travis, just like [coveralls.io](http://coveralls.io) is.

> Per-function performance can be measured by writing benchmark for this functions.

Yes, but you still get into the same run-to-run problems. :)

**Comment by sindresorhus** [comment]
> I tested it a little and yes - results differs from run to run, so benchmarks with slow operations will get false-negative from comparing benchmark results.

Run it multiple times in a new child process each time, remove the outlier and find the median.

I'm sure I've oversimplified it, but that's what I would start by doing.

**Comment by Qix-** [comment]
The problem with that is, depending on the load, travis will run your instance at a much lower priority to begin with, which means no matter how many times you run it, it's still going to be slower than past results.

**Comment by sindresorhus** [comment]
Good point. I'm sure there are some reading material or papers on the internet somewhere about this. We can't be the first wanting something like this. Maybe in other programming languages.

**Comment by Qix-** [comment]
There are; I'd have to dig them up though. IIRC there isn't an amazing way to calculate this.

**Comment by sindresorhus** [comment]
> IIRC there isn't an amazing way to calculate this.

Is there ever with computers :p

**Comment by Qix-** [comment]
Using giant calculators for reliable calculations? Never.

**Comment by sindresorhus** [comment]
What's the point of calculators, I just count on my fingers.

**Comment by Qix-** [comment]
*[stack based languages intensify]*

**Comment by Qix-** [comment]
@sindresorhus @floatdrop one of the ideas I had today was to benchmark a *standard* piece of code (i.e. 1,000,000 runs against a function call that sums `1` + `n`) and making it the *standard* unit of measurement. I haven't tested it on any platform, but if the highest priority is used then theoretically thread scheduling noise will be at the minimal possible level that we can achieve in user-space.

That value is then taken and applied to the rest of the values (probably `code-value / standard value`) to get the adjusted (normalized?) value.

I'd have to run some extensive tests, but it might be worth thinking about. We need a bridge to relate two relative values to an absolute. The function-sum-loop code, as mentioned earlier, is the *absolute*, whereas the performance is *relative* to the machine, though under the right circumstances we can, for the sake of ballpark-measurements, assume the deltas will be close to the same regardless of computer.

If any of that made sense.

---

For example,

```javascript
function double(n) {
  return n + n;
}

// time the following code, per iteration - a regular benchmark
var s = 0;
for (var i = 0; i < 1000000; i++) {
   s += double(i);
}

// i.e. it took 1000 milliseconds
```

The above is run for **every benchmark**, before the target program's code is executed. The averaged benchmarking time then becomes the *standard unit* used for that particular execution session.

Then, tests are run, i.e. `perfy _mocha`, and we get our results.

- The whole program took `2000m` to run. Its **overall score** is `2000 / si`, or `2000 / 1000`: **2**
- A function was just refactored and was expected to be faster (prior to refactoring, it took `6500ms` to run) but this time around it was actually `9000ms`.
  - Original score: `6500ms / si`: **6.5**
  - Refactored score: `9000ms / si`: **9.0**
  - Delta score: `9.0 - 6.5`: **+2.5** (this is bad)

The *score* isn't in seconds; if the control test took only `500ms`, and the above example tests took the same amount of time to run, then all of the example scores above would be halved.

---

I think this is the realistic best you can do in userspace.

**Comment by sindresorhus** [comment]
// @mathiasbynens @jdalton @logicalparadox

**Comment by jussi-kalliokoski** [comment]
[On twitter](https://twitter.com/quinnirill/status/635703769959931905) I tweeted a (tongue-in-cheek) idea that is quite extreme considering the workload, but basically the issue here is that time per CPU cycle is variable - this could be mitigated by running in an x86 emulator / VM that uses a fake clock that advances deterministically by each cycle. No trivial task though, more just an idea, since you'd have to emulate threads, sleep, etc.

**Comment by Qix-** [comment]
@jussi-kalliokoski the problem with that is still that the OS's thread scheduler will make the times variable. You might reduce the scheduling noise but unless you run Node *as the kernel itself*, you're going to have a degree of variance.

**Comment by aretecode** [comment]
I created this https://github.com/fluents/bench-chain, it has all the data required to show performance changes over time along with tags for naming those changes. It just needs some more ui to display that specific information in a more useful way.
